feat(service-worker): bypass on queryparam/header (#21191)

petersalomonsen opened the pull request:
issue #21191

added ngsw-bypass header and queryparameter
for bypassing serviceworker on specific requests

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 21191


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petersalomonsen commented:
@gkalpak I believe everything from #21191 is covered now. Let me know if anything else is needed for this Pull Request.

gkalpak commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -92,6 +92,15 @@ to refresh the resource in the background. This way, broken
 unhashed resources do not remain in the cache beyond their
 configured lifetimes.
 
+### Bypass the service worker
```
I would move this to the end of the "Service worker and caching of app resources" section.
I, also, think that "Bypassing" sounds more consistent with the other headings.

petersalomonsen replied:
This is done.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -176,6 +176,11 @@ export class Driver implements Debuggable, UpdateSource {
   private onFetch(event: FetchEvent): void {
     const req = event.request;
+
+    if (req.headers.has('ngsw-bypass') || req.url.indexOf('ngsw-bypass=true') > -1) {
```
1. I suggest handling both the header and the query param similarly: Either check for presense only or check for a value of `'true'`.
    One problem with requiring `'true'` is that it implies that adifferent value (e.g. `'false'`) disables the by-passing, which in turn raises the question of which values are valid, how to handle invalid values (if any), whether the header or the query param has precedence, etc. Therefore, I might lean towards checking for presense only (and making it clear in the docs).

2. The query param check needs to be stricter. E.g. extend `adapter.parseUrl()` to also return `.search` and check something like `requestUrlObj.search.slice(1).toLowerCase().split('&').includes('ngsw-bypass' /*or ngsw-bypass=true*/)`.

petersalomonsen replied:
Thanks for the review @gkalpak , and good feedback. I see your point, but just want to check with you before proceeding with stricter checks.

If acceptable I'd like to keep it as simple as possible (hopefully lower risk of bugs), and as CPU efficient as possible ( less expensive to search for a string than splitting, converting to lowercase, using regex etc ).

So I've modified the docs, so that it's clearer on the criteria for bypassing the SW. Now it says you can have either the request header with any value, or the query parameter with value `true`. I think adding the value requirement reduces the risks of collision with `ngsw-bypass` other places in the URL (like in the path, or other parameters with different postfix). Checking for `&` means that you can't have the parameter directly after `?` so I'd prefer not checking for `&`.

So my point here is that keeping it simple makes it easier for others to understand, and it's also faster code, while making it more complex also could make it more vulnerable and in that case more difficult to debug.

What do you think?

gkalpak replied:
I'm afraid I will have to disagree :grin:

Simpler code is great but only if the resulting requirements are simple. I would aim for simpler (and more intuitive) requirements first. I still stand by my earlier suggestion to:
1. Have consistency between header and query param values.
2. Correctly detect query param.
3. Be case insensitive (which again keeps consistency between header and query param).

To address some of your concerns directly:

> I'd like to keep it as simple as possible (hopefully lower risk of bugs) [...].

Keeping it too simple would indeed lower the risk of bugs in our code, but increase the risk of bugs in app code.

> I'd like to keep it [...] as CPU efficient as possible (less expensive to search for a string than splitting, converting to lowercase, using regex etc).

I wouldn't be so worried about computational complexity, because this is not a hot path and the other operations involved are much more expensive anyway.

petersalomonsen replied:
point taken :-) I'll try to implement it as you suggest. hopefully tomorrow.

petersalomonsen replied:
@gkalpak now only checking for presence of the `ngsw-bypass` query param, and it's case insensitive. I added the `search` property to the result of `adapter.parseUrl()` as you suggested.

Hopefully closer now to what you have in mind now, maybe you want some more test cases? I added some more.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -711,6 +711,24 @@ import {async_beforeEach, async_fit, async_it} from './async';
+
+      expect(await makeRequest(scope, '/bar.txt?ngsw-bypass=false')).toBe('this is bar');
+      server.assertSawRequestFor('/bar.txt');
+    });
```
Depending on how we decide to handle header/query param values, we need more checks to cover for other cases. E.g.:
- Lower/Upper/Mixed case.
- Falsy value/No value.
- Having more headers/query params.
- Having similarly named headers/query params (e.g. `?no-ngsw-bypass=true`).
- Having both the header and the query param.

gkalpak replied:
Thx for adding the extra tests. They look great.

Can you please add a couple more:
- Upper/Mixed case headers.
- Falsy value/No value (both headers and query param).

gkalpak replied:
I'd like to see one more test with an empty query param (e.g. `?ngsw-bypass=&foo=ngsw-bypass`).

gkalpak reviewed (changes requested):
Thx for taking this on, @petersalomonsen. I have left some minor comments, but generally this LGTM.
Also, can you please update the commit message to follow our [guidelines][1] more closely. E.g. something like:

> feat(service-worker): support bypassing SW with specific header/query param
>
> Add support for bypassing the ServiceWorker for a request by using the `ngsw-bypass` header or query
> parameter.
>
> Fixes #21191

[1]: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -176,9 +176,16 @@ export class Driver implements Debuggable, UpdateSource {
 
+    if (req.headers.has('ngsw-bypass') ||
+        requestUrlObj.search.slice(1).toLowerCase().split('&').find(
+            param => param.split('=')[0] === 'ngsw-bypass')) {
```
Since you are concerned about performance, it might be faster to avoid `splitting` each key-value pair (e.g. use `param === 'ngsw-bypass' || param.startsWith('ngsw-bypass=')`).

Or even using a RegExp (and avoiding all the slicing/lower-casing/splitting/searching in array): `/[?&]ngsw-bypass(?:[=&]|$)/i.test(requestUrlObj.search)`

**UPDATE:**
It turns out, avoiding splitting key-value pairs is twice as fast and using a RegExp is 7x faster.
(Again, not that all this should have any noticeable impact on the overall operation :grin:)

[1]: http://jsben.ch/iPz1G

petersalomonsen replied:
You are doing all the work here @gkalpak :-) I'm learning a lot, so very happy to be part of the process here! Very nice benchmark ( I suspected regexp was faster, but didn't really benchmark as you did ). I copied your RegExp which seems to do the job.

gkalpak commented on packages/service-worker/worker/testing/scope.ts:
```diff
@@ -184,14 +184,15 @@ export class SwTestHarness implements ServiceWorkerGlobalScope, Adapter, Context
+  parseUrl(url: string, relativeTo?: string): {origin: string, path: string, search: string} {
     const parsedUrl: URL = (typeof URL === 'function') ?
         new URL(url, relativeTo) :
         require('url').parse(require('url').resolve(relativeTo || '', url));
```
AFAICT, Node's `url.parse(...)` will set `.search` to `null` (instead of an empty string).
So, we either need to normalize `null` to an empty string or type the return value as `search: null | string`.

petersalomonsen replied:
I chose to return an empty string instead of changing the return type, since the use of `url.parse(...)` only applies to older nodejs versions (todays version has `URL` type).

gkalpak reviewed (changes requested):
Thx, @petersalomonsen. Looking much better already :wink:
I left some more suggestions (but we are close) 😇

gkalpak commented on packages/service-worker/worker/testing/scope.ts:
```diff
@@ -184,14 +184,15 @@ export class SwTestHarness implements ServiceWorkerGlobalScope, Adapter, Context
     return {
       origin: parsedUrl.origin || `${parsedUrl.protocol}//${parsedUrl.host}`,
       path: parsedUrl.pathname,
+      search: parsedUrl.search ? parsedUrl.search : ''
```
Nit: Could be simplified to `parsedUrl.search || ''`.

petersalomonsen replied:
Good point. Did that now.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -711,6 +711,72 @@ import {async_beforeEach, async_fit, async_it} from './async';
+      await makeRequest(scope, '/bar.txt?ngsw-bypaSS=something');
+      server.assertNoRequestFor('/bar.txt');
+
+      await makeRequest(scope, '/bar.txt?testparam=test&ngsw-bypASS=anything');
```
🙈

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -711,6 +711,72 @@ import {async_beforeEach, async_fit, async_it} from './async';
+      await makeRequest(scope, '/bar?testparam=test&ngsw-bypass&testparam2');
+      server.assertNoRequestFor('/bar');
+
+
```
Nit: Unnecessary newline.

petersalomonsen replied:
Fixed

gkalpak reviewed (approved):
A couple of minor comments. LGTM otherwise 👍

petersalomonsen commented:
Also added one more test with empty query param as you asked for.

gkalpak reviewed (approved):
Thx, @petersalomonsen 👍
I have added a fixup commit with some suggested fixes. LMK if it looks good to you, so I can go ahead and mark this for merging.

##
For posterity:

I was pondering whether we should also recognize an `x-`-prefixed header, since that seems to be a common requirement for non-standard headers. But the [current recommendation][1] is discouraging the `x-` prefix.

It is easy to address this (in a non-breaking way) later, so let's re-consider, if this turns out to be a problem. Another approach could be to have the SW remove the header or query param altogether, but that would make the requests differ depending on whether the SW is enabled/disabled. so probably not a good idea.

[1]: https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions#answer-3561399

petersalomonsen commented:
Looks good @gkalpak . Also tested building the serviceworker package and using the resulting `ngsw-worker.js` and it works as it should.

gkalpak commented:
**merge-assistance:** No idea what is causing the `ci/angular: size` check failure or how to fix.

IgorMinar reviewed (approved):
Lgtm. Thanks

gkalpak commented:
For posterity:

As @alexeagle's suggested, I have compared the `core/hello_world/bundle.br` artifacts from [master](https://circleci.com/gh/angular/angular/297646#artifacts/containers/0) and [this PR](https://circleci.com/gh/angular/angular/298314#artifacts/containers/0) and there are identical.

petersalomonsen committed:
feat(service-worker): support bypassing SW with specific header/query param

Add support for bypassing the ServiceWorker for a request by using the
ngsw-bypass header or query parameter.

Fixes #21191

gkalpak committed:
fixup! feat(service-worker): support bypassing SW with specific header/query param

gkalpak committed:
fixup! feat(service-worker): support bypassing SW with specific header/query param
